frontend/src/Frontend/UI/Widgets.hs

```diff
@@ -176,59 +177,104 @@ uiRealInputElement cfg = do
     inputElement $ cfg & initialAttributes %~
         (<> ("type" =: "number")) . addInputElementCls . addNoAutofillAttrs
 
+uiCorrectingInputElement
+  :: forall t m a er explanation. DomBuilder t m
+  => MonadFix m
+  => (Text -> Maybe a)
+  -> (a -> Maybe (a, explanation))
+  -> (a -> Text)
+  -> InputElementConfig er t (DomBuilderSpace m)
+  -> m (InputElement er (DomBuilderSpace m) t, Dynamic t (Maybe a), Event t (a, Maybe explanation))
+uiCorrectingInputElement parse sanitize render cfg = mdo
+  ie <- inputElement $ cfg & inputElementConfig_setValue %~ (\e -> leftmost [attemptCorrection e, forceCorrection inp])
+  let
+    inp = _inputElement_input ie
+    val = value ie
+
+    sanitization :: Functor f => f Text -> f (Maybe (a, Maybe explanation))
+    sanitization = fmap $ \i -> ffor (parse i) $ \p ->
+      case sanitize p of
+        Nothing -> (p, Nothing)
+        Just (a, x) -> (a, Just x)
```

Comment:
Depending on your taste for Functor for tuples. ;)
```
maybe (p, Nothing) (fmap Just) $ sanitize p
```

Reply:
I was tempted to do this, but `second` instead of `fmap`